src/input.rs

```diff
@@ -50,32 +52,84 @@ impl InputFile {
 				// file. However, I don't know why.
 				count_byte: len + std::mem::size_of::<u32>() + 16,
 				file_bytes,
+				file_version: *version,
 			}),
 			_ => Err(anyhow!("first frame is not a header")),
 		}
 	}
 
-	fn read_data(
-		&mut self,
-		length: usize,
-		read_attachment: bool,
-	) -> Result<Vec<u8>, anyhow::Error> {
+	fn read_decrypt_frame(&mut self) -> Result<Vec<u8>, anyhow::Error> {
+		let mut hmac = [0u8; crate::decrypter::LENGTH_HMAC];
+		let length: usize;
+		let mut data;
+
+		if self.file_version == 0 {
+			// frame length is unencrypted, so read it directly:
+			length = self
+				.reader
+				.read_u32::<byteorder::BigEndian>()
+				.unwrap()
+				.try_into()
+				.unwrap();
+
+			data = vec![0u8; length - crate::decrypter::LENGTH_HMAC];
+
+			// read data and decrypt
+			self.reader.read_exact(&mut data)?;
+			data = self.decrypter.decrypt(&mut data, true);
+		} else {
+			// first read frame length
+			let mut encrypted_len = vec![0u8; 4];
+			self.reader.read_exact(&mut encrypted_len)?;
+
+			// hmac will be updated with the encrypted_len data later
+			let decrypted_len = self.decrypter.decrypt(&encrypted_len, false);
+
+			length = byteorder::BigEndian::read_u32(&decrypted_len)
+				.try_into()
+				.unwrap();
+
+			hmac = [0u8; crate::decrypter::LENGTH_HMAC];
+			data = vec![0u8; 4 + length - crate::decrypter::LENGTH_HMAC];
+
+			// read data and decrypt
+			self.reader.read_exact(&mut data[4..])?;
+			data[0] = encrypted_len[0];
+			data[1] = encrypted_len[1];
+			data[2] = encrypted_len[2];
+			data[3] = encrypted_len[3];
+
+			data = self.decrypter.decrypt(&mut data, true);
+			data = data[4..].to_vec();
```

Comment:
This does some data copying, so could probably be implemented in a much more efficient manner!